citadel/docs/siteconfig.txt
  


Site Configuration
by IGnatius T Foobar 

    Once your Citadel server is up and running, the first thing you'll
want to do is customize and tune it.  This can be done from the
text-based client with the .Aide System configuration General command,
or from WebCit (if you have it installed) by clicking "Advanced Options"
followed by "Edit site-wide configuration."   This document will show the
text client being used, but the available set of options are the same
either way. 

    The first set of options deal with the identification of your system. 
  

 Lobby> . Aide System configuration General
 Node name [uncnsrd]: 
 Fully qualified domain name [uncnsrd.mt-kisco.ny.us]: 
 Human readable node name [Uncensored]: 
 Modem dialup number [US 914 244 3252]: 
 Geographic location of this system [Mount Kisco, NY]: 
 Name of system administrator [IGnatius T Foobar]: 
 Paginator prompt [<jinkies!  more text on the next screen!>]: 
  


For "node name" you should enter the "short" node name of your system.  It
is used in a number of places, most notably for routing of messages
across a Citadel network. Citadel will attempt to set the default value
to the unqualified host name of your computer. 

Then enter the fully-qualified domain name [FQDN] of your system.  If you
are not on the Internet, you can simply set it to the same as your
unqualified host name. Otherwise you should
set this value to the host name by which your system is most commonly known. 

The field "Human-readable node name" is also called the "node title" or
"organization name."  It is used primarily for display purposes.  Set it to 
the actual name of your system as you want it to appear. 

If you have a modem or bank of modems answering data calls for your
system, enter it in the field marked "Modem dialup number."  Otherwise
you may make something up.  This is used for interoperability with older
Citadel-86 systems which use a telephone number for node identification. 

"Geographic location of this system" is another display field.  Enter a 
city and state, or city and country. 

"Name of system administrator" is important!  Any user who logs on with 
the name you enter here will automatically be granted Aide privileges.
This is one of two ways for the system administrator to grant
himself/herself Aide access to the system when initially setting it up.  (The
other is simply to have the first account created on a new installation.) 
  

    The next set of options are your system's security settings.  Before
delving into the actual options, we should review the various access
levels available on the system.  Citadel has seven access levels: 

     0 (Deleted).  A user whose access level is set to 0 will
       automatically be deleted by the system. 
     1 (New User).   Users at this level may only read messages.  Entering
       messages is prohibited, except in the Mail> room, where a message
       to 'sysop' may be entered. 
     2 (Problem User).  Also known as "Twit." 
     3 (Local User).  May enter messages, except in rooms shared on a  
       Citadel network. 
     4 (Network User).  May enter messages in every accessible room. 
     5 (Preferred User).  Use of this level is up to the whim of the
       system administrator. 
     6 (Aide).  Access is granted to the administrative functions of the
       system.  (This access level may also be granted to a user only for
       a specific room, please see "Room Aide" for
       more information.) 


 Require registration for new users  [No]: No 
 Initial access level for new users [4]: 
 Access level required to create rooms [4]: 
 Automatically give room aide privs to a user who creates a private room [No]: No 
 Automatically move problem user messages to twit room  [Yes]: Yes 
 Name of twit room [Trashcan]: 
 Restrict Internet mail to only those with that privilege  [No]: No 
 Name of room to log pages [Page Log Scandal]: 
  


"Registration" refers to the process of a user entering various personal
contact information (real name, address, telephone number, etc.) into the
system.  When enabled, this information is stored as a vCard object on
the system in two places: the user's  My Citadel Config>  room, and in
the  Global Address Book> room.  (Note: the latter should be made private
on publicly-accessible systems, for obvious reasons.) 

If you answer Yes to "Require registration for new users" then each new
user, upon creating a new account, will immediately be entered into the
registration process. 

"Initial access level for new users" should be set to 1 (New User) if you  
would like to review each new user's registration info before granting
them higher access.  This would be done periodically with the <.A>ide
<V>alidate-new-users command.   If you do not require registration, you
should set the initial access level to 4 (Network User). 

"Access level required to create rooms" is up to you.  You might wish to
restrict the creation of new rooms only to Aides, or you might wish to
allow anyone to create a room.  The latter is one of the Citadel
culture's most long-standing traditions; the former may be appropriate if
users are abusing this privilege. 

You have the ability to "Automatically give room aide privs to a user who  
creates a private room."  If you answer Yes, then any user who creates a
guess-name, passworded, or invitation-only room will automatically become
the room aide, and will have access to a subset of the <.A>ide command
set while in that room.  If you would rather grant this permission  
manually, answer No. 

Another tradition in the Citadel culture is to refrain from deleting 
problem users, but instead to "twit" them (reduce their access level to 2
[Problem User]).   You can then "Automatically move problem user messages
to twit room" (answer Yes, then specify "Name of twit room" and remember
to create that room).  If you employ this logic, any user with level 2
(Problem User) access will continue to have access to the same set of
rooms, but all messages posted will automatically be routed to the
Trashcan (or whatever you call your twit room). 

If you have the Internet mail gateway installed, you have the option of
restricting its use on a user-by-user basis.  If you wish to do this,
answer Yes to "Restrict Internet mail to only those with that privilege." 

"Name of room to log pages" is where you can specify a room to which all  
pages (also called express messages or instant messages) will be logged.
You may wish to do this for security reasons. 
  

    The next set of options deals with the tuning of your system.  It is
safe to leave these untouched. 
  

 Server connection idle timeout (in seconds) [900]: 
 Maximum concurrent sessions [20]: 
 Maximum message length [2147483647]: 
 Minimum number of worker threads [5]: 
 Maximum number of worker threads [256]: 
 Server-to-server networking password [xxxxx]: 
   
  


The "Server connection idle timeout" is for the connection between client
and server software.  It is not an idle timer for the user interface.
900 seconds (15 minutes) is the default and a sane setting. 

"Maximum concurrent sessions" is the highest number of user sessions you
wish to allow on your system at any given time.  Citadel can scale to
hundreds of concurrent users, but if you have limited hardware or (more
likely) limited bandwidth, you might wish to set a maximum.  You can also
set it to zero for no limit. 

"Maximum message length" is just that.  This could be a good way to
prevent enormous multimedia files from finding their way into your
message base. 

The minimum and maximum number of worker threads can be tuned to your
liking.  Citadel will attempt to keep one worker thread running per
session, within these constraints.  You should be aware that due to the
use of the worker thread model, Citadel can handle a alarge number of
concurrent sessions with a much smaller thread pool.  If you don't know
the programming theory behind multithreaded servers, you should leave
these parameters alone. 

The "server-to-server networking password" may be set to some secret value
to allow authenticated networking between Citadel servers. 
  

    The final set of options deals with the purging or expiry of old
objects off the system, and the new smtp/pop settings. 
  
  

 POP3 server port (-1 to disable) [110]: 
 SMTP server port (-1 to disable) [25]: 
 Default user purge time (days) [61]: 
 Default room purge time (days) [21]: 
 System default message expire policy (? for list) [2]: 
 Keep how many messages online? [150]: 
 Save this configuration? Yes 
  


New settings to (.A)ide (S)ystem (G)eneral: 

POP/SMTP settings. Citadel is now capable of being it's own internet
e-mail server (smtp) or pop server. If you want that functionality, enter
in the correct ports. Normally this will be 110 for pop or 25 for smtp. More on this is in mailsetup.html. 

Any user who does not log in for the period specified in "Default user
purge time" will be deleted the next time a purge is run.  This setting
may be modified on a per-user basis. 

"Default room purge time" behaves the same way, and may also be modified
on a per-room basis. 

"System default message expire policy" defines the way in which old
messages are expired (purged) off the system.  You can specify any of: 

     Purge by age (specify in days) 
     Purge by message count in the room (specify number of messages) 
     Do not purge at all 

  Again, this setting may be overridden on a per-floor basis, and the
floor setting may be overridden on a per-room basis. 
  

    When you're done, enter "Yes" to confirm the changes, or "No" to abort. 


ctdlphp/testuser/config.txt
iconbar|ib_displayas=0,ib_logo=0,ib_summary=1,ib_inbox=1,ib_calendar=1,ib_contacts=1,ib_notes=1,ib
_tasks=1,ib_rooms=1,ib_users=2,ib_chat=1,ib_advanced=1,ib_logoff=0,ib_citadel=1
roomlistview|rooms
emptyfloors|yes
weekstart|0
use_sig|no
startpage|dotskip&room=_BASEROOM_
signature|
daystart|8
floordiv_expanded|
current_iconbar|0
calhourformat|24
dayend|17



ChangeLog
$Log$
Revision 503.0  2004/01/28 04:31:31  ajc
* THIS IS 5.03

Revision 502.4  2004/01/28 04:30:48  ajc
* Create Room can now offer calendar and address book rooms

Revision 502.3  2004/01/17 21:25:30  ajc
* Add 'base DN' and 'bind DN' and 'bind DN password' to siteconfig
* Split the config for LDAP Connector into its own panel

Revision 502.2  2004/01/17 04:32:54  ajc
* Add site config options for Citadel servers which support LDAP

Revision 502.1  2003/10/24 02:50:19  ajc
* Clicking "Enter Message" in a room with an Address Book view opens
  a blank vCard entry screen.

Revision 502.0  2003/10/23 03:16:50  ajc
* THIS IS 5.02

Revision 501.2  2003/10/23 03:07:54  ajc
* event.c: Found and removed another unquoted multi-line string literal
* webcit.h: changed an #ifdef HAVE_ICAL_H
  to #ifdef WEBCIT_WITH_CALENDAR_SERVICE to avoid false positives

Revision 501.1  2003/10/22 01:42:12  ajc
* iconbar.c: removed a stray character or two...

Revision 501.0  2003/10/21 04:26:37  ajc
* THIS IS 5.01

Revision 500.30  2003/10/21 04:26:27  ajc
* bumped the version string to 5.01

Revision 500.29  2003/10/17 03:08:28  ajc
* Finished prettying up the vCard display

Revision 500.28  2003/10/17 02:31:45  ajc
* Multi-line string literals are deprecated in GCC 3.1 unless each line
  is wrapped in quotes.  Performed this fix so it'll compile.
* setup.c: removed references to unused variable tempfile[]

Revision 500.27  2003/10/09 03:17:02  ajc
* Started prettying up the vCard display.

Revision 500.26  2003/10/08 02:33:49  ajc
* Begin work on better handling of vCard display

Revision 500.25  2003/09/04 03:41:27  ajc
* siteconfig.c: changed some config items to drop-down boxes

Revision 500.24  2003/08/27 04:26:54  ajc
* Split the site configuration screen up into four different screens,
  and added a meta-menu

Revision 500.23  2003/08/24 03:31:50  ajc
* Finished the setup program
* Bumped the internal version number to 5.01

Revision 500.22  2003/08/22 00:47:27  ajc
* setup is now working, but it still needs a way to detect a running WebCit

Revision 500.21  2003/08/22 00:25:40  ajc
* Added a skeleton "setup" program to the tree.  It does not work yet.

Revision 500.20  2003/08/20 03:48:55  ajc
* Added iconbar shortcut to Contacts room
* Minimum required Citadel is now 6.11

Revision 500.19  2003/08/20 03:07:25  ajc
* Address book "single card" view now suppresses "reading nn message" and
  instead shows the letter indices across the top
* Clicking on an address in a vCard now connects to WebCit mail instead
  of a browser mailto: link

Revision 500.18  2003/08/19 04:18:43  ajc
* Changed things around in the address book view.  Made it sorted and
  more clickable and better looking...

Revision 500.17  2003/08/12 18:54:03  error
* messages.c: fix off-by-one-error in selector bar

Revision 500.16  2003/08/07 22:18:03  ajc
* Add the Summary Page to the customizable iconbar

Revision 500.15  2003/08/07 03:51:51  ajc
* Customizable icon bar as: pictures and text, pictures, text

Revision 500.14  2003/08/05 03:52:14  ajc
* First working version of the customizable iconbar.  It contains a basic
  set of options and now automatically refreshes when it's customized.

Revision 500.13  2003/08/04 03:50:29  ajc
* First cut at the customizable iconbar is done.  A bunch of generic options
  are there.  Still need to get the actual iconbar to refresh when changes
  are made.

Revision 500.12  2003/08/03 21:02:01  ajc
* More work on customizable iconbar

Revision 500.11  2003/08/02 05:25:46  ajc
* Laid the groundwork for a customizable iconbar

Revision 500.10  2003/07/30 03:57:49  ajc
* Add config.c_purge_hour to config screen

Revision 500.9  2003/07/06 22:06:59  ajc
* When converting "anything that looks like a URL" to a real link, first
  make sure that it isn't already inside a link.  This fixes most of the
  long, ugly URL strings sent by eBay.
* Automatically dismantle mailto: links in HTML messages, and convert
  them to WebCit mail links (clicking on it sends the user to the Mail
  room and begins composing a new message with the recipient and subject
  fields pre-populated).

Revision 500.8  2003/06/29 20:51:21  ajc
* Worked around a rendering bug in Satan's Browser that was causing it
  to draw the calendar month view much wider than the actual screen wdith

Revision 500.7  2003/06/28 04:09:37  error
* roomops.c: Backout change to display one less box in the first column in
  roomlist view, it doesn't seem to work for everyone

Revision 500.6  2003/06/28 03:52:51  error
* Don't tell browsers not to cache static documents

Revision 500.5  2003/06/28 03:35:23  error
* roomops.c: Display one less floor in the first column in romlist view

Revision 500.4  2003/06/26 03:37:58  ajc
* Allow anonymous HTTP requests for /freebusy/user%20name.vcf (or .vfb)
  This will allow some groupware clients to browse the free/busy times of
  other users on the system.

Revision 500.3  2003/06/21 05:17:21  ajc
* Better alignment of system messages
* Exterminated display_error() and replaced it with WC->ImportantMessage
  which is displayed in a bubble on top of whatever page you're on next.
* Small cosmetic changes
* Fixed some nesting problems in output_headers()

Revision 500.2  2003/06/20 03:39:27  ajc
* Optimization to floor layout in box view

Revision 500.1  2003/06/20 03:04:26  ajc
* "Reply" button on messages now handles IGnet recipients correctly

Revision 500.0  2003/06/10 04:23:08  ajc
* THIS IS 5.00

Revision 410.50  2003/06/10 04:22:58  ajc
* Updated docs and headers for 5.00 release

Revision 410.49  2003/06/10 04:05:23  ajc
* De-tabled the left side icon bar and reduced one of the graphics, so
  that all of the rounded corners show up in Satan's Browser.

Revision 410.48  2003/06/10 03:47:41  ajc
* style.css: added font selection to default <BODY> tag

Revision 410.47  2003/06/06 03:26:03  ajc
* Satan's Browser insists on putting bloody acres of whitespace to the

citadel/config.c
/*
 * Read and write the citadel.config file
 *
 * Copyright (c) 1987-2012 by the citadel.org team
 *
 * This program is open source software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

#include "sysdep.h"
#include <stdlib.h>
#include <sys/stat.h>
#include <unistd.h>
#include <stdio.h>
#include <errno.h>
#include <string.h>
#include <limits.h>
#include <sys/utsname.h>
#include <libcitadel.h>
#include "citadel.h"
#include "server.h"
#include "config.h"
#include "ctdl_module.h"

struct config config;

/*
 * Put some sane default values into our configuration.  Some will be overridden when we run setup.
 */
void brand_new_installation_set_defaults(void) {

	struct passwd *pw;
	struct utsname my_utsname;
	struct hostent *he;

	/* Determine our host name, in case we need to use it as a default */
	uname(&my_utsname);

	/* set some sample/default values in place of blanks... */
	char c_nodename[256];
	safestrncpy(c_nodename, my_utsname.nodename, sizeof c_nodename);
	strtok(config.c_nodename, ".");
	if (IsEmptyStr(config.c_fqdn) ) {
		if ((he = gethostbyname(my_utsname.nodename)) != NULL) {
			safestrncpy(config.c_fqdn, he->h_name, sizeof config.c_fqdn);
		}
		else {
			safestrncpy(config.c_fqdn, my_utsname.nodename, sizeof config.c_fqdn);
		}
	}

	safestrncpy(config.c_humannode, "Citadel Server", sizeof config.c_humannode);
	safestrncpy(config.c_phonenum, "US 800 555 1212", sizeof config.c_phonenum);
	config.c_initax = 4;
	safestrncpy(config.c_moreprompt, "<more>", sizeof config.c_moreprompt);
	safestrncpy(config.c_twitroom, "Trashcan", sizeof config.c_twitroom);
	safestrncpy(config.c_baseroom, BASEROOM, sizeof config.c_baseroom);
	safestrncpy(config.c_aideroom, "Aide", sizeof config.c_aideroom);
	config.c_port_number = 504;
	config.c_sleeping = 900;
	config.c_instant_expunge = 1;

	if (config.c_ctdluid == 0) {
		pw = getpwnam("citadel");
		if (pw != NULL) {
			config.c_ctdluid = pw->pw_uid;
		}
	}
	if (config.c_ctdluid == 0) {
		pw = getpwnam("bbs");
		if (pw != NULL) {
			config.c_ctdluid = pw->pw_uid;
		}
	}
	if (config.c_ctdluid == 0) {
		pw = getpwnam("guest");
		if (pw != NULL) {
			config.c_ctdluid = pw->pw_uid;
		}
	}
	if (config.c_createax == 0) {
		config.c_createax = 3;
	}

	/*
	 * Default port numbers for various services
	 */
	config.c_smtp_port = 25;
	config.c_pop3_port = 110;
	config.c_imap_port = 143;
	config.c_msa_port = 587;
	config.c_smtps_port = 465;
	config.c_pop3s_port = 995;
	config.c_imaps_port = 993;
	config.c_pftcpdict_port = -1 ;
	config.c_managesieve_port = 2020;
	config.c_xmpp_c2s_port = 5222;
	config.c_xmpp_s2s_port = 5269;
}



/*
 * get_config() is called during the initialization of Citadel server.
 * It verifies the system's integrity and reads citadel.config into memory.
 */
void get_config(void) {
	FILE *cfp;
	int rv;

	if (chdir(ctdl_bbsbase_dir) != 0) {
		fprintf(stderr,
			"This program could not be started.\nUnable to change directory to %s\nError: %s\n",
			ctdl_bbsbase_dir,
			strerror(errno)
		);
		exit(CTDLEXIT_HOME);
	}

	memset(&config, 0, sizeof(struct config));
	cfp = fopen(file_citadel_config, "rb");
	if (cfp != NULL) {
		rv = fread((char *) &config, sizeof(struct config), 1, cfp);
		if (rv != 1)
		{
			fprintf(stderr, 
				"Warning: The config file %s has unexpected size. \n",
				file_citadel_config);
		}
		fclose(cfp);
	}
	else {
		brand_new_installation_set_defaults();
	}

	/* Ensure that we are linked to the correct version of libcitadel */
	if (libcitadel_version_number() < LIBCITADEL_VERSION_NUMBER) {
		fprintf(stderr, "    You are running libcitadel version %d.%02d\n",
			(libcitadel_version_number() / 100), (libcitadel_version_number() % 100));
		fprintf(stderr, "citserver was compiled against version %d.%02d\n",
			(LIBCITADEL_VERSION_NUMBER / 100), (LIBCITADEL_VERSION_NUMBER % 100));
		exit(CTDLEXIT_LIBCITADEL);
	}

	/* Only allow LDAP auth mode if we actually have LDAP support */
#ifndef HAVE_LDAP
	if ((config.c_auth_mode == AUTHMODE_LDAP) || (config.c_auth_mode == AUTHMODE_LDAP_AD)) {
		fprintf(stderr, "Your system is configured for LDAP authentication,\n"
				"but you are running a server built without OpenLDAP support.\n");
		exit(CTDL_EXIT_UNSUP_AUTH);
	}
#endif

        /* Default maximum message length is 10 megabytes.  This is site
	 * configurable.  Also check to make sure the limit has not been
	 * set below 8192 bytes.
         */
        if (config.c_maxmsglen <= 0)
                config.c_maxmsglen = 10485760;
        if (config.c_maxmsglen < 8192)
                config.c_maxmsglen = 8192;

        /* Default lower and upper limits on number of worker threads */

	if (config.c_min_workers < 3)		/* no less than 3 */
		config.c_min_workers = 5;

	if (config.c_max_workers == 0)			/* default maximum */
		config.c_max_workers = 256;

	if (config.c_max_workers < config.c_min_workers)   /* max >= min */
		config.c_max_workers = config.c_min_workers;

	/* Networking more than once every five minutes just isn't sane */
	if (config.c_net_freq == 0L)
		config.c_net_freq = 3600L;	/* once per hour default */
	if (config.c_net_freq < 300L) 
		config.c_net_freq = 300L;

	/* Same goes for POP3 */
	if (config.c_pop3_fetch == 0L)
		config.c_pop3_fetch = 3600L;	/* once per hour default */
	if (config.c_pop3_fetch < 300L) 
		config.c_pop3_fetch = 300L;
	if (config.c_pop3_fastest == 0L)
		config.c_pop3_fastest = 3600L;	/* once per hour default */
	if (config.c_pop3_fastest < 300L) 
		config.c_pop3_fastest = 300L;

	/* "create new user" only works with native authentication mode */
	if (config.c_auth_mode != AUTHMODE_NATIVE) {
		config.c_disable_newu = 1;
	}
}


/*
 * Occasionally, we will need to write the config file, because some operations
 * change site-wide parameters.
 */
void put_config(void)
{
	FILE *cfp;
	int rv;

	if ((cfp = fopen(file_citadel_config, "rb+")) == NULL)
		perror(file_citadel_config);
	else {
		rv = fwrite((char *) &config, sizeof(struct config), 1, cfp);
		if (rv == -1)
			syslog(LOG_EMERG, "Failed to write: %s [%s]\n", 
			       file_citadel_config, strerror(errno));
		fclose(cfp);
	}
}
